src/lib/consensus/proof_of_stake.ml

```diff
@@ -1022,38 +1020,96 @@ module Make (Inputs : Inputs_intf) :
+                ( Length.pack_var previous_state.curr_epoch_data.length
+                  :> Field.var )
+          in
+          Length.var_of_field Field.(Checked.(add (constant one) base))
+        and ledger =
+          Epoch_ledger.if_ epoch_increased
+            ~then_:
+              { total_currency= previous_state.total_currency
+              ; hash= previous_blockchain_state_ledger_hash }
+            ~else_:previous_state.curr_epoch_data.ledger
+        and start_checkpoint =
+          Coda_base.State_hash.if_ epoch_increased
+            ~then_:previous_protocol_state_hash
+            ~else_:previous_state.curr_epoch_data.start_checkpoint
+        (* Want this to be the protocol state hash once we leave the seed
+           update range. *)
+        and lock_checkpoint =
+          let%bind base =
+            (* TODO: Should this be zero or some other sentinel value? *)
+            Coda_base.State_hash.if_ epoch_increased
+              ~then_:Coda_base.State_hash.(var_of_t (of_hash zero))
+              ~else_:previous_state.curr_epoch_data.lock_checkpoint
+          in
+          let%bind in_seed_update_range =
+            Epoch.Slot.in_seed_update_range_var previous_state.curr_slot
+          in
+          Coda_base.State_hash.if_ in_seed_update_range
+            ~then_:previous_protocol_state_hash ~else_:base
+        in
+        {Epoch_data.seed; length; ledger; start_checkpoint; lock_checkpoint}
+      and length = Length.increment_var previous_state.length
       (* TODO: keep track of total_currency in transaction snark. The current_slot
        * implementation would allow an adversary to make then total_currency incorrect by
        * not adding the coinbase to their account. *)
-      and total_currency =
-        Amount.Checked.add previous_state.total_currency
-          (Amount.var_of_t Inputs.coinbase)
-      in
-      let%bind total_currency =
-        Amount.Checked.add total_currency supply_increase
+      and new_total_currency =
+        Amount.Checked.add previous_state.total_currency supply_increase
+      and epoch_length =
+        Length.increment_if_var previous_state.epoch_length epoch_increased
       in
-      (* TODO: check vrf result from transition data *)
-      let%map last_epoch_data, curr_epoch_data, epoch_length =
-        Epoch_data.update_pair_checked
-          (previous_state.last_epoch_data, previous_state.curr_epoch_data)
-          previous_state.epoch_length ~prev_epoch:previous_state.curr_epoch
-          ~next_epoch:transition_data.epoch ~curr_slot:previous_state.curr_slot
-          ~prev_protocol_state_hash:previous_protocol_state_hash
-          ~proposer_vrf_result:transition_data.proposer_vrf_result ~ledger_hash
-          ~total_currency
-      in
-      { length
-      ; epoch_length
-      ; curr_epoch= transition_data.epoch
-      ; curr_slot= transition_data.slot
-      ; total_currency
-      ; last_epoch_data
-      ; curr_epoch_data }
+      return
```

Comment:
Would be nice to see this conform to the original function pattern of the ocaml implementation. Alternatively, if you don't agree with/like the way the current ocaml implementation, you could change that as well, but imo they should be roughly equivalent in layout for debugging and spot checking.